test/unit/logger/logger.cpp

```diff
@@ -113,9 +114,98 @@ TEST_F(LoggerWithFileSink, NestedFilePath) {
     std::filesystem::create_directories(file_path);
     file_path /= file_name;
     logger = std::make_unique<logger::Logger>(
-        logger::Level::WARN,
-        std::make_unique<logger::FileSink>(logger_name, file_path));
+        logger::Level::WARN, std::make_unique<logger::FileSink>(
+                                 logger_name, file_path, logger::Level::WARN));
 
     logger->warning("Test message: {}", "success");
-    test_msg += "[WARNING]: Test message: success\n";
+    test_msg << test_msg_prefix << "[WARNING]: Test message: success\n";
+}
+
+//////////////////////////////////////////////////////////////////////////////
+INSTANTIATE_TEST_SUITE_P(
+    ThreadCount, FileSinkLoggerMultipleThreads,
+    ::testing::Values(2 * std::thread::hardware_concurrency(),
+                      std::thread::hardware_concurrency(),
+                      std::thread::hardware_concurrency() / 2));
+TEST_P(FileSinkLoggerMultipleThreads, Multithreaded) {
+    std::vector<std::thread> threads;
+    auto local_logger = logger::Logger(
+        logger::Level::WARN,
+        std::make_unique<logger::FileSink>(logger_name, file_path, true));
+    constexpr int message_count = 50;
+
+    // Messages below the flush level
+    for (int i = 0; i < thread_count; i++) {
+        threads.emplace_back([&]() {
+            for (int j = 0; j < message_count; ++j) {
+                local_logger.warn("Test message: {}", "it's a success");
+            }
+        });
+    }
+
+    for (auto &thread : threads) {
+        thread.join();
+    }
+    threads.clear();
+
+    // Messages at the flush level
+    for (int i = 0; i < thread_count; i++) {
+        threads.emplace_back([&]() {
+            for (int j = 0; j < message_count; ++j) {
+                local_logger.error("Flushed test message: {}",
+                                   "it's a success");
+            }
+        });
+    }
+
+    for (auto &thread : threads) {
+        thread.join();
+    }
+
+    for (int i = 0; i < thread_count * message_count; ++i) {
+        test_msg << "Test message: it's a success\n";
+    }
+    for (int i = 0; i < thread_count * message_count; ++i) {
+        test_msg << "Flushed test message: it's a success\n";
+    }
+}
+
+//////////////////////////////////////////////////////////////////////////////
+INSTANTIATE_TEST_SUITE_P(
+    ThreadCount, CommonLoggerWithMultipleThreads,
+    ::testing::Values(2, std::thread::hardware_concurrency()));
+TEST_P(CommonLoggerWithMultipleThreads, StdoutMultithreaded) {
+    std::vector<std::thread> threads;
+    auto local_logger =
+        logger::Logger(logger::Level::WARN,
+                       std::make_unique<logger::StdoutSink>("test", true));
+    constexpr int message_count = 50;
```

Comment:
We could make the `message_count` a test suite parameter like the `thread_count`.